### PR TITLE
Much faster shortening and expanding via direct GMP usage

### PR DESCRIPTION
**Pull request by Ocramius**
Turns out that this library was a massive bottleneck in a downstream application that I was working on: ~270 shortening operations already account for multiple seconds of processing time.

In order to speed up things, I wrote a benchmark suite and added a new `GMPShortener` that only works when `ext-gmp` is available.

TODOs:

 * [x] test any new added details
 * [x] add a factory to decide whether to instantiate a `Shortener` or a `GMPShortener` based on currently installed extensions
 * [x] update documentation

Performance difference thus far:

```
./vendor/bin/phpbench run --revs=5 --iterations=5 
PhpBench @git_tag@. Running benchmarks.
Using configuration file: /home/ocramius/Documents/keiko/uuid-shortener/phpbench.json

\Benchmark\Keiko\Uuid\Shortener\GMPShortenerBench

    benchShorteningOfTinyUuid...............R2 I4 [μ Mo]/r: 5.440 5.478 (μs) [μSD μRSD]/r: 0.150μs 2.75%
    benchShorteningOfHugeUuid...............R1 I1 [μ Mo]/r: 38.040 38.454 (μs) [μSD μRSD]/r: 0.731μs 1.92%
    benchShorteningOfPromiscuousUuids.......R2 I4 [μ Mo]/r: 3,888.760 3,896.839 (μs) [μSD μRSD]/r: 48.471μs 1.25%
    benchExpandingOfTinyUuid................R2 I2 [μ Mo]/r: 7.760 7.829 (μs) [μSD μRSD]/r: 0.196μs 2.53%
    benchExpandingOfHugeUuid................R1 I0 [μ Mo]/r: 26.600 26.376 (μs) [μSD μRSD]/r: 0.551μs 2.07%
    benchExpandingOfPromiscuousUuids........R1 I3 [μ Mo]/r: 2,519.760 2,503.429 (μs) [μSD μRSD]/r: 57.113μs 2.27%

\Benchmark\Keiko\Uuid\Shortener\ShortenerBench

    benchShorteningOfTinyUuid...............R3 I4 [μ Mo]/r: 32.920 32.433 (μs) [μSD μRSD]/r: 0.744μs 2.26%
    benchShorteningOfHugeUuid...............R1 I0 [μ Mo]/r: 564.600 560.265 (μs) [μSD μRSD]/r: 12.444μs 2.20%
    benchShorteningOfPromiscuousUuids.......R1 I4 [μ Mo]/r: 60,155.480 59,641.523 (μs) [μSD μRSD]/r: 1,061.669μs 1.76%
    benchExpandingOfTinyUuid................R1 I2 [μ Mo]/r: 30.120 29.815 (μs) [μSD μRSD]/r: 0.700μs 2.32%
    benchExpandingOfHugeUuid................R1 I0 [μ Mo]/r: 276.840 281.003 (μs) [μSD μRSD]/r: 6.780μs 2.45%
    benchExpandingOfPromiscuousUuids........R2 I4 [μ Mo]/r: 28,829.720 29,435.657 (μs) [μSD μRSD]/r: 861.582μs 2.99%

12 subjects, 60 iterations, 60 revs, 0 rejects, 0 failures, 0 warnings
(best [mean mode] worst) = 5.200 [8,031.337 8,038.258] 5.600 (μs)
⅀T: 481,880.200μs μSD/r 170.928μs μRSD/r: 2.231%
```

The new implementation is almost 20 times faster on a pre-defined set of 100 UUIDs to process.

### CODE REVIEW

File: src/GMPShortener.php [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        return $output;
+    }
+
+    public function expand(string $shortUuid): string
+    {
+        $number = gmp_init(0, 10);
+        $dictionaryLength = gmp_init($this->dictionary->length, 10);
+
+        foreach (str_split(strrev($shortUuid)) as $char) {
+            $number = gmp_add(gmp_mul($number, $dictionaryLength), (string) $this->dictionary->charIndexes[$char]);
```
Comment (Ocramius) [inline src/GMPShortener.php]: Hmm, this syntax will lead to a notice when no character like this exists: that's an un-tested edge case.

File: src/Shortener.php [hunk 2]
```diff
@@ -52,19 +68,20 @@ public function expand(string $shortUuid): string
         $number = BigInteger::zero();
         foreach (str_split(strrev($shortUuid)) as $char) {
             $number = $number
-                ->multipliedBy($this->dictionary->getLength())
-                ->plus($this->dictionary->getCharIndex($char));
+                ->multipliedBy($this->dictionary->length)
+                ->plus($this->dictionary->charIndexes[$char]);
```
Comment (Ocramius) [inline src/Shortener.php]: This syntax will lead to a notice when no character like this exists: that's an un-tested edge case.

Comment (Ocramius) [reply src/Shortener.php]: Fixed by ddb205b

File: src/GMPShortener.php (hunk 1 above)
Comment (Ocramius) [reply src/GMPShortener.php]: Fixed by ddb205b

**Review by mgrajcarek — APPROVED** [review]
@Ocramius another great PR!
Thank you very much!

**Comment by Ocramius** [comment]
Tanks for the quick merge, @mgrajcarek!

I've further improved this in #12: had missed that this is just a base conversion algorithm, that GMP already handles internally :D

**Pull request merged**